# Basic Gates Kata Workbook

**What is this workbook?** 
A workbook is a collection of problems, accompanied by solutions to them. The explanations focus on the logical steps required to solve a problem; they illustrate the concepts that need to be applied to come up with a solution to the problem, explaining the mathematical steps required.

Note that a workbook should not be the primary source of knowledge on the subject matter; it assumes that you've already read a tutorial or a textbook and that you are now seeking to improve your problem-solving skills. You should attempt solving the tasks of the respective kata first, and turn to the workbook only if stuck or for reinforcement. While a textbook emphasizes knowledge acquisition, a workbook emphasizes skill acquisition.

This workbook describes the solutions to the problems offered in the [Basic Gates Kata](./BasicGates.ipynb). 
Since the tasks are offered as programming problems, the explanations also cover some elements of Q# that might be non-obvious for a novitiate.

**What you should know for this workbook**

You should be familiar with the following concepts and associated techniques **prior to** beginning work on the Basic Gates Quantum Kata.

1. [Complex numbers](../tutorials/ComplexArithmetic/ComplexArithmetic.ipynb).
2. Basic linear algebra (multiplying column vectors by matrices), per the first part of [this tutorial](../tutorials/LinearAlgebra/LinearAlgebra.ipynb).
3. [The concept of qubit and its properties](../tutorials/Qubit/Qubit.ipynb).
4. [Single-qubit gates](../tutorials/SingleQubitGates/SingleQubitGates.ipynb).

You can also consult the [complete Quantum Katas learning path](https://github.com/microsoft/QuantumKatas#learning-path).

# Part 1. Single-Qubit Gates


## Task 1.1. State flip: $|0\rangle$ to $|1\rangle$ and vice versa

**Input:** A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal:**  Change the state of the qubit to $\alpha |1\rangle + \beta |0\rangle$.

**Example:**

If the qubit is in state $|0\rangle$, change its state to $|1\rangle$.

If the qubit is in state $|1\rangle$, change its state to $|0\rangle$.

### Solution

We can recognise that the Pauli X gate will change the state $|0\rangle$ to $|1\rangle$ and vice versa, and $\alpha |0\rangle + \beta |1\rangle$ to $\alpha |1\rangle + \beta |0\rangle$.

As a reminder, the Pauli X gate is defined by the following matrix:

$$
X = 
\begin{bmatrix}
   0 &  1\\
  1 &  0
\end{bmatrix}
$$

We can see how it affects, for example, the basis state $|0\rangle$:

$$X|0\rangle= 
\begin{bmatrix}
   0 &  1\\
  1 &  0
\end{bmatrix}
\begin{bmatrix}
   1\\
   0
\end{bmatrix}
=
\begin{bmatrix}
   0 \cdot 1 + 1 \cdot 0\\
   1 \cdot 1 + 0 \cdot 0
\end{bmatrix}
=
\begin{bmatrix}
   0\\
   1
\end{bmatrix}
=|1\rangle
$$

Similarly, we can consider the effect of the X gate on the superposition state $|\psi\rangle = 0.6|0\rangle + 0.8|1\rangle$:

$$X|\psi\rangle= 
\begin{bmatrix}
   0 &  1\\
  1 &  0
\end{bmatrix}
\begin{bmatrix}
   0.6\\
   0.8
\end{bmatrix}
=
\begin{bmatrix}
  0 \cdot 0.6  + 1 \cdot 0.8\\
   1 \cdot 0.6 + 0 \cdot 0.8
\end{bmatrix}
=
\begin{bmatrix}
   0.8\\
   0.6
\end{bmatrix}
=  0.8|0\rangle + 0.6|1\rangle
$$

In [1]:
%kata T101_StateFlip 

operation StateFlip (q : Qubit) : Unit is Adj+Ctl {
    X(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.8 + 0𝑖
|1⟩	0.6 + 0𝑖

The actual state:


|0⟩	0.8 + 0𝑖
|1⟩	0.6 + 0𝑖

Success!

[Return to Task 1.1 of the Basic Gates kata.](./BasicGates.ipynb#Task-1.1.-State-flip:-$|0\rangle$-to-$|1\rangle$-and-vice-versa)

## Task 1.2. Basis change: $|0\rangle$ to $|+\rangle$ and $|1\rangle$ to $|-\rangle$ (and vice versa)

**Input**: A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal**:  Change the state of the qubit as follows:
* If the qubit is in state $|0\rangle$, change its state to $|+\rangle = \frac{1}{\sqrt{2}} \big(|0\rangle + |1\rangle\big)$.
* If the qubit is in state $|1\rangle$, change its state to $|-\rangle = \frac{1}{\sqrt{2}} \big(|0\rangle - |1\rangle\big)$.
* If the qubit is in superposition, change its state according to the effect on basis vectors.

### Solution

We can recognize that the Hadamard gate changes states $|0\rangle$ and $|1\rangle$ to $|+\rangle$ and $|-\rangle$, respectively, and vice versa.

As a reminder, the Hadamard gate is defined by the following matrix:

$$
\frac{1}{\sqrt{2}}\begin{bmatrix}
1 & 1 \\
1 & -1
\end{bmatrix}
$$

For example, we can work out $H|1\rangle$ as follows:

$$
H|1\rangle=
\frac{1}{\sqrt{2}}\begin{bmatrix}
   1 & 1 \\
   1 & -1
  \end{bmatrix}
 \begin{bmatrix}
   0\\
   1\\
  \end{bmatrix}
=
\frac{1}{\sqrt{2}}\begin{bmatrix}
   1 \cdot 0 + 1 \cdot 1 \\
   1 \cdot 0 + (-1) \cdot 1
  \end{bmatrix}
=
  \frac{1}{\sqrt{2}}\begin{bmatrix}
   1\\
   -1
  \end{bmatrix}
= \frac{1}{\sqrt{2}} \big(|0\rangle - |1\rangle\big) = |-\rangle
$$

Similarly, we can consider the effect of the Hadamard gate on the superposition state $|\psi\rangle = 0.6|0\rangle + 0.8|1\rangle$ (rounding the numbers to 4 decimal places):

$$
H|\psi⟩ = 
\frac{1}{\sqrt{2}}\begin{bmatrix}
   1 & 1 \\
   1 & -1
  \end{bmatrix}
 \begin{bmatrix}
   \alpha\\
   \beta\\
  \end{bmatrix}
 =
\frac{1}{\sqrt{2}}\begin{bmatrix}
   \alpha + \beta\\
   \alpha - \beta\\
  \end{bmatrix}
= 0.7071\begin{bmatrix}
   1.4\\
   -0.2\\
  \end{bmatrix}
= \begin{bmatrix}
   0.98994\\
   -0.14142\\
  \end{bmatrix}
= 0.9899|0\rangle - 0.1414|1\rangle
 $$

In [2]:
%kata T102_BasisChange 

operation BasisChange (q : Qubit) : Unit is Adj+Ctl {
    H(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.9899494936611666 + 0𝑖
|1⟩	-0.1414213562373096 + 0𝑖

The actual state:


|0⟩	0.9899494936611665 + 0𝑖
|1⟩	-0.1414213562373096 + 0𝑖

Success!

[Return to Task 1.2 of the Basic Gates kata](./BasicGates.ipynb#Task-1.2.-Basis-change:-$|0\rangle$-to-$|+\rangle$-and-$|1\rangle$-to-$|-\rangle$-(and-vice-versa)).

## Task 1.3. Sign flip: $|+\rangle$  to $|-\rangle$  and vice versa.

**Input**: A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal** :  Change the qubit state to $\alpha |0\rangle - \beta |1\rangle$ (i.e. flip the sign of the $|1\rangle$ component of the superposition).


### Solution

The action of the Pauli Z gate is exactly what is required by this question.
This gate leaves the sign of the $|0\rangle$ component of the superposition unchanged but flips the sign of the $|1\rangle$ component of the superposition.

As a reminder, the Pauli Z gate is defined by the following matrix:

$$
Z = 
 \begin{bmatrix}
   1 & 0\\
   0 & -1
  \end{bmatrix}
 $$
 
Let's see its effect on the only computational basis state that it changes, $|1\rangle$:

$$
Z|1\rangle =
 \begin{bmatrix}
   1 & 0\\
   0 & -1
  \end{bmatrix}
 \begin{bmatrix}
   0\\
   1\\
  \end{bmatrix}
=
\begin{bmatrix}
   1 \cdot 0 + 0 \cdot1\\
   0 \cdot 1 +  -1 \cdot 1\\
  \end{bmatrix}
=
 \begin{bmatrix}
   0\\
   -1\\
  \end{bmatrix}
=
 -\begin{bmatrix}
   0\\
   1\\
  \end{bmatrix}
= -|1\rangle
$$

In general applying the Z gate to a single qubit superposition state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$ gives

$$
Z|\psi\rangle = 
 \begin{bmatrix}
   1 & 0 \\
   0 & -1
  \end{bmatrix}
 \begin{bmatrix}
   \alpha\\
   \beta\\
  \end{bmatrix}
  =
\begin{bmatrix}
   1\cdot\alpha + 0\cdot\beta\\
   0\cdot\alpha + -1\cdot\beta\\
  \end{bmatrix} 
  =
 \begin{bmatrix}
   \alpha\\
   -\beta\\
  \end{bmatrix}  
= \alpha |0\rangle -\beta |1\rangle
$$

In [3]:
%kata T103_SignFlip

operation SignFlip (q : Qubit) : Unit is Adj+Ctl {
    Z(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.6 + 0𝑖
|1⟩	-0.8 + 0𝑖

The actual state:


|0⟩	-0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

Success!

[Return to Task 1.3 of the Basic Gates kata](./BasicGates.ipynb#Task-1.3.-Sign-flip:-$|+\rangle$--to-$|-\rangle$--and-vice-versa.).

## Task 1.4. Amplitude change: $|0\rangle$ to $\cos{α} |0\rangle + \sin{α} |1\rangle$.

**Inputs:**

1. Angle α, in radians, represented as Double.
2. A qubit in state $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$.

**Goal:**  Change the state of the qubit as follows:
- If the qubit is in state $|0\rangle$, change its state to $\cos{α} |0\rangle + \sin{α} |1\rangle$.
- If the qubit is in state $|1\rangle$, change its state to $-\sin{α} |0\rangle + \cos{α} |1\rangle$.
- If the qubit is in superposition, change its state according to the effect on basis vectors.

### Solution

We can recognise that we need to use one of the rotation gates Rx, Ry, and Rz (named because they "rotate" the qubit state in the three dimensional space visualized as the Bloch sphere about the x, y, and z axes, respectively), since they involve angle parameters. Of these three gates, only Ry rotates the basis states $|0\rangle$ and $|1\rangle$ to have real amplitudes (the other two gates introduce complex coefficients).

As a reminder, 

$$
R_{y}(\theta) =
\begin{bmatrix}
   \cos \frac{\theta}{2} & -\sin \frac{\theta}{2}\\
   \sin \frac{\theta}{2} & \cos \frac{\theta}{2}
  \end{bmatrix}
$$

Let's see its effect on the $|0\rangle$ state:

$$
R_y(\theta)|0\rangle =
\begin{bmatrix}
   \cos \frac{\theta}{2} & -\sin \frac{\theta}{2}\\
   \sin \frac{\theta}{2} & \cos \frac{\theta}{2}
  \end{bmatrix}
\begin{bmatrix}
   1\\
   0\\
\end{bmatrix}
=
\begin{bmatrix}
   \cos \frac{\theta}{2}\cdot1 - \sin \frac{\theta}{2}\cdot0\\
   \sin \frac{\theta}{2}\cdot1 + \cos \frac{\theta}{2}\cdot0
  \end{bmatrix}
=
\begin{bmatrix}
   \cos \frac{\theta}{2}\\
   \sin \frac{\theta}{2}
  \end{bmatrix}
= \cos\frac{\theta}{2} |0\rangle + \sin\frac{\theta}{2} |1\rangle
$$

Recall that when applying a gate, you can tell what its matrix does to the basis states by looking at its columns: the first column of the matrix is the state into which it will transform the $|0\rangle$ state, and the second column is the state into which it will transfrom the $|1\rangle$ state.

In the example used by the testing harness we are given $\beta = 0.6, \gamma = 0.8$ and $\alpha = 1.0471975511965976 = \frac{\pi}{3}$.

Since $\cos \frac{\pi}{3} = 0.5$  and $\sin \frac{\pi}{3} = 0.8660$, working to 4 decimal places, we can compute:

$$
R_{y}(\theta) |\psi\rangle
=
 \begin{bmatrix}
   \cos \frac{\theta}{2} & -\sin \frac{\theta}{2}\\
   \sin \frac{\theta}{2} & \cos \frac{\theta}{2}
  \end{bmatrix} 
  \begin{bmatrix}
   \beta\\
  \gamma
 \end{bmatrix}
=
\begin{bmatrix}
   cos \frac{\theta}{2}\cdot\beta - sin \frac{\theta}{2}\cdot\gamma\\
   sin \frac{\theta}{2}\cdot\beta +cos \frac{\theta}{2}\cdot\gamma
  \end{bmatrix}
=  
 \begin{bmatrix}
   0.6\cdot\cos \frac{\pi}{3} -0.8\cdot\sin \frac{\pi}{3}\\
   0.6\cdot\sin \frac{\pi}{3} +0.8\cdot\cos \frac{\pi}{3}
  \end{bmatrix}
=
 \begin{bmatrix}
   0.3 - 0.6928\\
   0.5196 + 0.4
  \end{bmatrix}
=
  \begin{bmatrix}
   -0.3928\\
    0.9196
  \end{bmatrix}
$$ 

Notice that we used $\frac{\theta}{2} = \alpha$; this means that in the Q# code we need to pass the angle $\theta = 2\alpha$.

In [4]:
%kata T104_AmplitudeChange

operation AmplitudeChange (alpha : Double, q : Qubit) : Unit is Adj+Ctl {
    Ry(2.0 * alpha, q);
}

Applying amplitude change with alpha = 1.0471975511965976
The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	-0.3928203230275509 + 0𝑖
|1⟩	0.9196152422706633 + 0𝑖

The actual state:


|0⟩	-0.3928203230275509 + 0𝑖
|1⟩	0.9196152422706633 + 0𝑖

Success!

[Return to Task 1.4 of the Basic Gates kata](./BasicGates.ipynb#Task-1.4.-Amplitude-change:-$|0\rangle$-to-$\cos{α}-|0\rangle-+-\sin{α}-|1\rangle$.).

## Task 1.5. Phase flip

**Input:** A qubit in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

**Goal:** Change the qubit state to $\alpha |0\rangle + \color{red}i\beta |1\rangle$ (add a relative phase $i$ to $|1\rangle$ component of the superposition).


### Solution

We can recognise that the S gate performs this particular relative phase addition to the $|1\rangle$ basis state. As a reminder,

$$
S = 
 \begin{bmatrix}
   1 & 0\\
   0 & i
  \end{bmatrix}
 $$

Let's see the effect of this gate on the general superposition $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$.

$$
 \begin{bmatrix}
   1 & 0 \\
   0 & i
  \end{bmatrix}
 \begin{bmatrix}
   \alpha\\
   \beta\\
  \end{bmatrix}
  =
\begin{bmatrix}
   1\cdot\alpha + 0\cdot\beta\\
   0\cdot\alpha + i\cdot\beta
  \end{bmatrix} 
  =
 \begin{bmatrix}
   \alpha\\
   i\beta\\
  \end{bmatrix}  
$$

It is therefore easy to see that when $|\psi\rangle = 0.6|0\rangle +  0.8|1\rangle, S|\psi\rangle =  0.6|0\rangle + 0.8i|1\rangle$.

In [5]:
%kata T105_PhaseFlip

operation PhaseFlip (q : Qubit) : Unit is Adj+Ctl {
    S(q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.6 + 0𝑖
|1⟩	0 + 0.8𝑖

The actual state:


|0⟩	0 + 0.6𝑖
|1⟩	-0.8 + 0𝑖

Success!

### Solution 2

See the next task, Phase Change, for an explanation of using R1 gate to implement the same transformation:

In [6]:
%kata T105_PhaseFlip

open Microsoft.Quantum.Math;

operation PhaseFlip (q : Qubit) : Unit is Adj+Ctl {
    R1(0.5 * PI(), q);
}

The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.6 + 0𝑖
|1⟩	0 + 0.8𝑖

The actual state:


|0⟩	0 + 0.6𝑖
|1⟩	-0.8 + 1.1102230246251565E-16𝑖

Success!

[Return to Task 1.5 of the Basic Gates kata](./BasicGates.ipynb#Task-1.5.-Phase-flip).

## Task 1.6. Phase change

**Inputs:**

1. Angle α, in radians, represented as Double.
2. A qubit in state $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$.

**Goal:**  Change the state of the qubit as follows:
- If the qubit is in state $|0\rangle$, don't change its state.
- If the qubit is in state $|1\rangle$, change its state to $e^{i\alpha} |1\rangle$.
- If the qubit is in superposition, change its state according to the effect on basis vectors: $\beta |0\rangle + \color{red}{e^{i\alpha}} \gamma |1\rangle$.

### Solution
We know that:

$$
R1(\alpha)
= 
 \begin{bmatrix}
   1 & 0\\
   0 & \color{red}{e^{i\alpha}}
  \end{bmatrix}
 $$

So we have:

$$
R1(\beta |0\rangle + \gamma |1\rangle) = 
 \begin{bmatrix}
   1 & 0 \\
   0 & \color{red}{e^{i\alpha}}
  \end{bmatrix}
 \begin{bmatrix}
   \beta\\
   \gamma\\
  \end{bmatrix}
  =
\begin{bmatrix}
   1.\beta + 0.\gamma\\
   0.\beta + \color{red}{e^{i\alpha}}\gamma
  \end{bmatrix} 
  =
 \begin{bmatrix}
   \beta\\
   \color{red}{e^{i\alpha}}\gamma
  \end{bmatrix}  
  = \beta |0\rangle + \color{red}{e^{i\alpha}} \gamma |1\rangle
$$

> Note that the results produced by the test harness can be unexpected.
If you run the kata several times and  and examine the output, you'll notice that success is signalled even though the corresponding amplitudes of the desired and actual states look very different.
>
> So what's going on? The full state simulator used in these tests performs the computations "up to a global phase", that is, sometimes the resulting state acquires a global phase that doesn't affect the computations or the measurement outcomes, but shows up in DumpMachine output. (You can read more about the global phase in the [Qubit tutorial](../tutorials/Qubit/Qubit.ipynb#Relative-and-Global-Phase).)
>
> For example, in one run you can get the desired state $(0.6000 + 0000i)|0\rangle + (-0.1389 +0.7878i)|1\rangle$ and the actual state $(-0.1042 + 0.5909i)|0\rangle + (-0.7518 -0.2736i)|1\rangle$.
You can verify that the ratios of amplitudes of the respective basis states are equal: $\frac{-0.1042 + 0.5909i}{0.6} = -0.173667 +0.984833 i = \frac{-0.7518 -0.2736i}{-0.1389 +0.7878i}$, so the global phase acquired by the state is (-0.173667 +0.984833 i). 
You can also check that the absolute value of this multiplier is approximately 1, so it doesn't impact the measurement probabilities.
>
> The testing harness for this and the rest of the tasks checks that your solution implements the required transformation exactly, without introducing any global phase, so it shows up only in the helper output and does not affect the verification of your solution.

In [7]:
%kata T106_PhaseChange

operation PhaseChange (alpha : Double, q : Qubit) : Unit is Adj+Ctl {
   R1(alpha, q);
}

Applying phase change with alpha = 1.7453292519943295
The starting state:


|0⟩	0.6 + 0𝑖
|1⟩	0.8 + 0𝑖

The desired state:


|0⟩	0.6 + 0𝑖
|1⟩	-0.13891854213354424 + 0.7878462024097665𝑖

The actual state:


|0⟩	-0.10418890660015818 + 0.5908846518073249𝑖
|1⟩	-0.7517540966287269 + -0.2736161146605349𝑖

Success!

Suppose now that $\alpha = \frac{\pi}{2}$.
Then $e^{i\alpha}= \cos\frac{\pi}{2} + i\sin\frac{\pi}{2}$.
And, since $\cos\frac{\pi}{2}= 0$ and $\sin\frac{\pi}{2} = 1$, then we have that $\cos\frac{\pi}{2} + i \sin\frac{\pi}{2} = i$, and  
$R1(\frac{\pi}{2}) = S$, which we used in the second solution to task 1.5, above.

[Return to Task 1.6 of the Basic Gates kata](./BasicGates.ipynb#Task-1.6.-Phase-Change).

## Task 1.7. Global phase change
**Input:** A qubit in state $|\psi\rangle = \beta |0\rangle + \gamma |1\rangle$.

**Goal**: Change the state of the qubit to $- \beta |0\rangle - \gamma |1\rangle$.

> Note: this change on its own is not observable - there is no experiment you can do on a standalone qubit to figure out whether it acquired the global phase or not. 
> However, you can use a controlled version of this operation to observe the global phase it introduces. 
> This is used in later katas as part of more complicated tasks.

### Solution

We recognise that a global phase change can be accomplished by using the R rotation gate with the PauliI (identity) gate.
As a reminder, the R gate is defined as $R_{\mu}(\theta) = \exp(\frac{\theta}{2}i\cdot\sigma_{\mu})$, wehere $\sigma_{\mu}$ is one of the Pauli gates I, X, Y or Z.  

> Note that a global phase is not detectable and has no physical meaning - it disappers when you take a measurement of the state.  
> You can read more about this in the [Single-qubit measurements tutorial](../tutorials/SingleQubitSystemMeasurements/SingleQubitSystemMeasurements.ipynb#Measurements-in-arbitrary-orthogonal-bases).

For the problem at hand, we'll use the rotation gate $R_{\mu}(\theta) = \exp(\frac{\theta}{2}i\cdot\sigma_{\mu})$ with $\sigma_{\mu} = I$.  

$R(PauliI, 2\pi) = \exp(\frac{2\pi}{2} iI) = \exp(i\pi) I = (\cos\pi + i\sin\pi) I$ and, since $\cos\pi = -1$   and   $\sin\pi = 0$, we have that $R(PauliI, 2\pi) = -I$:   

$$
R(\beta |0\rangle + \gamma |1\rangle) 
=
 -1\begin{bmatrix}
   1 & 0 \\
   0 & 1
  \end{bmatrix}
 \begin{bmatrix}
   \beta\\
   \gamma\\
  \end{bmatrix}
=
 \begin{bmatrix}
   -1 & 0 \\
   0 & -1
  \end{bmatrix}
 \begin{bmatrix}
   \beta\\
   \gamma\\
  \end{bmatrix} 
=
 \begin{bmatrix}
   -1\cdot\beta + 0\cdot\gamma\\
    0\cdot\beta +  -1\cdot\gamma \\
  \end{bmatrix}
=
\begin{bmatrix}
   -\beta\\
   -\gamma\\
  \end{bmatrix}
=
- \beta |0\rangle - \gamma |1\rangle
$$

The test harness for this test shows the result of applying the *controlled* variant of your solution to be able to detect the phase change.

In [8]:
%kata T107_GlobalPhaseChange

open Microsoft.Quantum.Math;

operation GlobalPhaseChange (q : Qubit) : Unit is Adj+Ctl {
    R(PauliI, 2.0 * PI(), q);
}

Showing effect of controlled-GlobalPhaseChange
The starting state:


|00⟩	0.4242640687119285 + 0𝑖
|01⟩	0.5656854249492381 + 0𝑖
|10⟩	0.4242640687119285 + 0𝑖
|11⟩	0.5656854249492381 + 0𝑖

The desired state:


|00⟩	0.4242640687119285 + 0𝑖
|01⟩	0.5656854249492381 + 0𝑖
|10⟩	-0.4242640687119285 + -5.19573633741296E-17𝑖
|11⟩	-0.5656854249492381 + -6.927648449883948E-17𝑖

The actual state:


|00⟩	0.4242640687119285 + 0𝑖
|01⟩	0.5656854249492381 + 0𝑖
|10⟩	-0.4242640687119285 + -5.19573633741296E-17𝑖
|11⟩	-0.5656854249492381 + -6.927648449883948E-17𝑖

Success!

[Return to Task 1.7 of the Basic Gates kata](./BasicGates.ipynb#Task-1.7.-Global-phase-change).

## Task 1.8. Bell state change - 1

**Input:** Two entangled qubits in Bell state $|\Phi^{+}\rangle = \frac{1}{\sqrt{2}} \big(|00\rangle + |11\rangle\big)$.

**Goal:**  Change the two-qubit state to $|\Phi^{-}\rangle = \frac{1}{\sqrt{2}} \big(|00\rangle - |11\rangle\big)$.

### Solution

We recognise that the goal is another Bell state. In fact, it is one of the four Bell states.

We remember from Task 1.3 that the Pauli Z gate will change the state of the $|0\rangle$ basis state of a single qubit, so this gate seems like a good candidate for what we want to achieve. This gate leaves the sign of the $|0\rangle$ basis state of a superposition unchanged, but flips the sign of the $|1\rangle$ basis state of the superposition.  

Don't forget that the Z gate acts on only a single qubit, and we have two here.
Lets also remember how the Bell state is made up from its individual qubits.

If the two qubits are A and B, where A is `qs[0]` and and B is `qs[1]`, we can write that
$|\Phi^{+}\rangle = \frac{1}{\sqrt{2}} \big(|0_{A}0_{B}\rangle + |1_{A}1_{B}\rangle\big)$,  and also remember that this is a superposition.

If we apply the Z gate to the qubit A, it will flip the phase of the basis state $|1_A\rangle$. As this phase is in a sense spread across the entangled state, with $|1_A\rangle$ basis state being part of the second half of the superposition, this  application has the effect of flipping the sign of the whole basis state $|1_A1_B\rangle$, as you can see by running the solution below.  

> If you run the solution below several times, you will notice that you get the state $\frac{1}{\sqrt{2}} \big(-|00\rangle + |11\rangle\big)$ at a guess as frequently as the goal state. Here is why.  
> 
> We can write this second state as   $-\frac{1}{\sqrt{2}} \big(|00\rangle -|11\rangle\big)$ = $(-1)\frac{1}{\sqrt{2}} \big(|00\rangle - |11\rangle\big)$. 
This state is the same as our goal state $|\Phi^{-}\rangle$ with an unmeasureable global phase of $-1$.

The exact same calculations can be done if we apply Z to the qubit B, so that's another possible solution.

In [9]:
%kata T108_BellStateChange1

operation BellStateChange1 (qs : Qubit[]) : Unit is Adj+Ctl {
    Z(qs[0]); // or Z(qs[1]); 
}

The starting state:


|00⟩	0.7071067811865476 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0 + 0𝑖
|11⟩	0.7071067811865476 + 0𝑖

The desired state:


|00⟩	0.7071067811865476 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0 + 0𝑖
|11⟩	-0.7071067811865476 + 0𝑖

The actual state:


|00⟩	-0.7071067811865476 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0 + 0𝑖
|11⟩	0.7071067811865476 + 0𝑖

Success!

[Return to Task 1.8 of the Basic Gates kata](./BasicGates.ipynb#Task-1.8.-Bell-state-change---1).

# Task 1.9. Bell state change - 2

**Input:** Two entangled qubits in Bell state $|\Phi^{+}\rangle = \frac{1}{\sqrt{2}} \big(|00\rangle + |11\rangle\big)$.

**Goal:**  Change the two-qubit state to $|\Psi^{+}\rangle = \frac{1}{\sqrt{2}} \big(|01\rangle + |10\rangle\big)$.

### Solution ##
We have seen in Task 1.1 that the Pauli X gate flips $|0\rangle$ to $|1\rangle$ and vice versa, and as we seem to need some flipping of states, perhaps this gate may be of use. (Bearing in mind, of course, that the X gate operates on a single qubit). 

Let's compare the starting state $\frac{1}{\sqrt{2}} \big(|0_A0_B\rangle + |1_A1_B\rangle\big)$ with the goal state $\frac{1}{\sqrt{2}} \big(1_A0_B\rangle + |0_A1_B\rangle\big)$ term by term and see how we need to transform it to reach the goal.

Using our nomencalure from Task 1.8, we can now see by comparing terms that $|0_{A}\rangle$ has flipped to $|1_A\rangle$ to get the first term, and $|1_{A}\rangle$ has flipped to $|0\rangle$ to get the second term. This allows us to say that the correct gate to use is Pauli X, applied to `qs[0]`.

In [10]:
%kata T109_BellStateChange2

operation BellStateChange2 (qs : Qubit[]) : Unit is Adj+Ctl {
    X(qs[0]); // or X(qs[1]);
}

The starting state:


|00⟩	0.7071067811865476 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0 + 0𝑖
|11⟩	0.7071067811865476 + 0𝑖

The desired state:


|00⟩	0 + 0𝑖
|01⟩	0.7071067811865476 + 0𝑖
|10⟩	0.7071067811865476 + 0𝑖
|11⟩	0 + 0𝑖

The actual state:


|00⟩	0 + 0𝑖
|01⟩	0.7071067811865476 + 0𝑖
|10⟩	0.7071067811865476 + 0𝑖
|11⟩	0 + 0𝑖

Success!

[Return to Task 1.9 of the Basic Gates kata](./BasicGates.ipynb#Task-1.9.-Bell-state-change---2).

## Task 1.10. Bell state change - 3

**Input:** Two entangled qubits in Bell state $|\Phi^{+}\rangle = \frac{1}{\sqrt{2}} \big(|00\rangle + |11\rangle\big)$.

**Goal:**  Change the two-qubit state, without adding a global phase, to $|\Psi^{-}\rangle = \frac{1}{\sqrt{2}} \big(|01\rangle - |10\rangle\big)$.

## Solution ##
We remember from Task 1.3 that the Pauli-Z gate leaves the sign of the $|0\rangle$ component of the single qubit superposition unchanged but flips the sign of the $|1\rangle$ component of the superposition. We have also just seen in Task 1.9 how to change our input state to the state $\frac{1}{\sqrt{2}} \big(|01\rangle + |10\rangle\big)$, which is almost our goal state (disregarding the phase change for the moment). So it would seem that a combination of these two gates wiill be what we need here. The remaining question is in what order to apply them, and to which qubit.

First of all, which qubit? Looking back at Task 1.9, it seems clear that we need to use qubit `qs[0]`, like we did there.

Second, in what order should we apply the gates? Remember that the Pauli-Z gate flips the phase of the $|1\rangle$ component of the superposition and leaves the $|0\rangle$ component alone.   
Let's experiment with applying X to `qs[0]` first. Looking at our "halfway answer" state $\frac{1}{\sqrt{2}} \big(|01\rangle + |10\rangle\big)$, we can see that if we apply the Z gate to `qs[0]`, it will leave the $|0_{A}\rangle$ alone but flip the phase of $|1_{A}\rangle$ to $-|1_{A}\rangle$, thus flipping the phase of the $|11\rangle$ component of our  Bell state.  

In [11]:
%kata T110_BellStateChange3

operation BellStateChange3 (qs : Qubit[]) : Unit is Adj+Ctl {
    X(qs[0]);
    Z(qs[0]);
}

The starting state:


|00⟩	0.7071067811865476 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0 + 0𝑖
|11⟩	0.7071067811865476 + 0𝑖

The desired state:


|00⟩	0 + 0𝑖
|01⟩	0.7071067811865476 + 0𝑖
|10⟩	-0.7071067811865476 + 0𝑖
|11⟩	0 + 0𝑖

The actual state:


|00⟩	0 + 0𝑖
|01⟩	-0.7071067811865476 + 0𝑖
|10⟩	0.7071067811865476 + 0𝑖
|11⟩	0 + 0𝑖

If the desired and the actual states match but the test doesn't pass, check whether your solution introduces a global phase; it shouldn't!


Success!

[Return to Task 1.10 of the Basic Gates kata](./BasicGates.ipynb#Task-1.10.-Bell-state-change---3).

## Part II. Multi-Qubit Gates

### Task 2.1. Two-qubit gate - 1

## Solution

**Input:** Two unentangled qubits (stored in an array of length 2).
The first qubit will be in state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$, the second - in state $|0\rangle$
(this can be written as two-qubit state  $\big(\alpha |0\rangle + \beta |1\rangle \big) \otimes |0\rangle = \alpha |00\rangle + \beta |10\rangle$.


**Goal:**  Change the two-qubit state to $\alpha |00\rangle + \beta |11\rangle$.

> Note that unless the starting state of the first qubit was $|0\rangle$ or $|1\rangle$,
> the resulting two-qubit state can not be represented as a tensor product
> of the states of individual qubits any longer; thus the qubits become entangled.

For our Task, lets use $A= \alpha |0\rangle + \beta |1\rangle$  and  $B= |0\rangle$

Compare our two inputs qubits in the state $|\psi\rangle = \alpha |0\rangle + \beta |1\rangle$ with the output qubits in the state $\alpha |00\rangle + \beta |11\rangle$. We want to pass our inputs through a gate, or gates, to be decided, that do the following. If A is in a 0 state ,then we want to leave B alone. However if A is in 1 state we want to flip B from 0into a 1 state. So, in other words, the state of B is to be made contingent upon the state of A. This gate exists and is called a CNOT (for controlled not)  gate.Dependent upon the value of what is know as the **control** qubit (A in our case) the value of the controlled or **target** qubit (B in our case) is inverted.  Thus we get the output of $\alpha |00\rangle + \beta |11\rangle$.

It's worth emphasizing that A is initially a superposition, having a measured value of either  $|\alpha^{2}|$   or  $|\beta^{2}|$ where $|\alpha^{2}| + |\beta^{2}|$ for normalisation.

It's worth remembering the significance of  Bell pairs, refering to the $|\Phi^{+}\rangle$ pair, given as the goal, in this case. The goal can be described, as we have done in Task 1.8, as $|\Phi^{+}\rangle = \alpha |0_{A}0_{B}\rangle + \beta|1_{A}1_{B}\rangle$.  
If A measures her qubit and gets the state $|0\rangle$, (with a probability of  $|\alpha^{2}|$) and if B **then** measures his qubit he will also get a $|0\rangle$  because the whole state has 'collapsed' under the measurement of A's qubit leaving B with  $|0\rangle$. We could represent this measurement as $\alpha |00\rangle + \beta |10\rangle -->|00\rangle$.  Similarly, A measuring a $|1\rangle$ will guarantee that B measures the same. An identical situation obtains if B measures first, guaranteeing  A the same result as B. It is important to keep well in mind when thinking about and computing with Bell states that the way state is written doesn't mean or imply that the qubits are physically joined in any way; they can be any arbitrary distance apart and the Bell state property we have described has beeen confirmed experimentally many times.

In [12]:
%kata T201_TwoQubitGate1
operation TwoQubitGate1(qs : Qubit[]) : Unit is Adj+Ctl {
        CNOT(qs[0], qs[1]);
    }

The starting state:


|00⟩	0.5000000000000001 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0.8660254037844386 + 0𝑖
|11⟩	0 + 0𝑖

The desired state:


|00⟩	0.5000000000000001 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0 + 0𝑖
|11⟩	0.8660254037844386 + 0𝑖

The actual state:


|00⟩	0.5000000000000001 + 0𝑖
|01⟩	0 + 0𝑖
|10⟩	0 + 0𝑖
|11⟩	0.8660254037844386 + 0𝑖

Success!

[Return to Task 2.1. of the Basic Gates kata](./BasicGates.ipynb#Task-2.1.-Two-qubit-gate---1).